### ISSUE TITLE
t3470-multithread-reactor-run.py issue test intermittent failures

### ISSUE THREAD

**Issue by garlick** [flux-framework/flux-core#3723]
This test is failing intermittently on my system (Ubuntu focal, `python-3.8.2-0ubuntu2`), more often when running tests in parallel but occasionally standalone.
 
```
$ ./t4000-issues-test-driver.t -v
[snip]
expecting success: run_timeout 120 /home/garlick/proj/flux-core/t/issues/t3470-multithread-reactor-run.py
starting 2 threads
got response from <Thread(Thread-1, started 140374035314432)>
got response from <Thread(Thread-2, started 140373959636736)>
2 threads started
published test-event
1: hello
Fatal Python error: ffi.from_handle() detected that the address passed points to garbage. If it is really the result of ffi.new_handle(), then the Python object has already been garbage collected
Python runtime state: initialized

Current thread 0x00007fab6082b700 (most recent call first):
  File "/home/garlick/proj/flux-core/src/bindings/python/flux/message.py", line 122 in message_handler_wrapper
  File "/home/garlick/proj/flux-core/src/bindings/python/flux/wrapper.py", line 194 in __call__
  File "/home/garlick/proj/flux-core/src/bindings/python/flux/core/handle.py", line 274 in reactor_run
  File "/home/garlick/proj/flux-core/t/issues/t3470-multithread-reactor-run.py", line 23 in get_events
  File "/usr/lib/python3.8/threading.py", line 870 in run
  File "/usr/lib/python3.8/threading.py", line 932 in _bootstrap_inner
  File "/usr/lib/python3.8/threading.py", line 890 in _bootstrap

Thread 0x00007fab63f42740 (most recent call first):
  File "/usr/lib/python3.8/threading.py", line 1027 in _wait_for_tstate_lock
  File "/usr/lib/python3.8/threading.py", line 1011 in join
  File "/home/garlick/proj/flux-core/t/issues/t3470-multithread-reactor-run.py", line 55 in main
  File "/home/garlick/proj/flux-core/t/issues/t3470-multithread-reactor-run.py", line 61 in <module>
not ok 15 - t3470-multithread-reactor-run.py
#	run_timeout 120 /home/garlick/proj/flux-core/t/issues/t3470-multithread-reactor-run.py
[snip]
```

**Comment by garlick** [comment]
Is it possible that the watcher is eligible for garbage collection immediately upon being started, since it is not (directly)? referenced from the `get_events()` thread after that?

I have been running `t4000-issues-test-driver.t` in a loop with the change below for a while on a loaded system, and haven't yet seen a failure:
```diff
diff --git a/t/issues/t3470-multithread-reactor-run.py b/t/issues/t3470-multithread-reactor-run.py
index efcb6fb19..c1fd5b5f2 100755
--- a/t/issues/t3470-multithread-reactor-run.py
+++ b/t/issues/t3470-multithread-reactor-run.py
@@ -19,8 +19,10 @@ def get_events(i, queue):
     f = flux.Flux()
     f.event_subscribe("test-event")
     queue.put(True)
-    f.msg_watcher_create(cb, topic_glob="test-event", args=i).start()
+    w = f.msg_watcher_create(cb, topic_glob="test-event", args=i)
+    w.start()
     f.reactor_run()
+    w.destroy()
 
 
 def main():
```

**Comment by grondo** [comment]
Oh, heh. Nice catch! It is beyond me, but your fix here makes sense.

Perhaps, like futures, we need a table in the reactor class or object to capture "active" watchers so they are not garbage collected out from under a running reactor loop?

**Comment by garlick** [comment]
Thanks - I'll open a separate issue on that and submit this patch, if that's OK.  (I'm a bit out of my depth in the bindings internals)
